kotlinx-coroutines-core/common/src/flow/DataFlow.kt

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow
+
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.flow.internal.*
+import kotlinx.coroutines.internal.*
+import kotlin.coroutines.*
+import kotlin.jvm.*
+
+/**
+ * A [Flow] that contains a single updatable data [value].
+ *
+ * A data flow can be created with `DataFlow()` constructor function either without an initial value
+ * or with an initial value. Data value can be updated by setting its [value] property.
+ * Updates to the [value] are always [conflated][Flow.conflate]. So a slow collector will skip fast updates,
+ * but always collects the most recently emitted value.
+ *
+ * The name reflects the fact that [DataFlow] represents an cell in a "data flow programming" model
+ * (think of an electronic spreadsheet). Dependent values can be defined by using various operators on the
+ * flows, with [combineLatest] being especially useful to combine multiple data flows using arbitrary functions.
+ *
+ * A data flow can be [closed][close] and all its collectors will complete.
+ *
+ * A read-only interface to data flow is a [Flow]. It is supposed to be collected by the consumers of the data.
+ * The ability to read the current [value] is provided only for convenience of the code that updates the data.
```

Comment:
Is CAS not supported by design? The `CounterModel` example is not _thread-safe_.

Reply:
Yes. It is assume that the owner of `LiveData` that makes updates is single-threaded. Do you have compelling use-cases to support CAS there to allow coordinate updates from multiple threads?

Reply:
I often use a `Channel` or a `Mutex` to resolve this kind of issues, so these use cases can be resolved outside the `DataFlow` class.

`Flow` is a safe abstraction, but the example `counter.value++` is not. I am not proposing a some kind of CAS operation, however there is not possible to declare that `counter.value++` simply works.

Similar to other @qwwdfsad's considerations, we can consider `DataFlow<T> : Flow<T>, FlowCollector<T>`, this adds a lot of _cerimonies_ but it make evident any issue in `counter.emit( counter.first() + 1)`.